More features for PossessionMod

xAstroBoy opened the pull request:
Add various features such as :

Enemy automatically opening internal doors with cooldown.
Enemy able to exit /enter from the exit /enter doors with cooldown
add a calibrate collision and a way to preserve the enemy colliders as well
Made the speed sync along with the enemy's animated speed.

Added Del key to delete enemy under possession  if needed (useful for scares)
Added SpringMan AI overrides

added various enemy events into controller.

added a controllable Interact radius based off which enemy.

Current Denied enemies for entrances are Forest Giant and leviatan worm and for oblivious reasons the snare flea as well.

During possession if you interact with a enemy and use the doors to exit, it will as well update the enemy's navigation system so it will still work even when you leave the possession

added isAbleToRotate as well to lock camera rotation while enemy is doing special animations.

winstxnhdw commented on lc-hax/Scripts/Modules/Possession/PossessionMod.cs:
```diff
@@ -190,15 +233,110 @@ internal sealed class PossessionMod : MonoBehaviour {
+        Transform? exitPoint = this.GetExitPointFromDoor(teleport);
+        if (exitPoint == null) return;
+        Enemy.isOutside = !teleport.isEntranceToBuilding;
+        Enemy.allAINodes = Enemy.isOutside ? GameObject.FindGameObjectsWithTag("OutsideAINode") : GameObject.FindGameObjectsWithTag("AINode");
```
Oh this is interesting. Did you find this with UnityExplorer?

xAstroBoy replied:
A little bit of both, i saw how lethal escape works and i saw how it sets the outside nodes

xAstroBoy commented:
@winstxnhdw just added BlobAI as well

D1GQ commented:
The thumper and sometimes Flower Man movement is very jittery I think it has something to do with their animation and their speed interpretating from their animations.

xAstroBoy commented:
> The thumper and sometimes Flower Man movement is very jittery I think it has something to do with their animation and their speed interpretating from their animations.

giant as well, ill add a CanSyncSpeed and set these 3 entities to false

xAstroBoy commented:
@D1GQ @winstxnhdw 

i finished a bunch of changes

Add Crawler controller
TestEnemy

add Speed Sync Option

Block Bracken, Giant, Leviatan , Snare , Crawler animation speed sync
Add TestEnemy and Crawler controllers

Add The possiblity to make the Hoarding Bug Angry as well.

add SetOutsideStatus to update the Nodes of a enemy instance.

Add F9 to restore Possessed Enemy ' s AI control while possessing it.

Fix Hoarding bug easily getting angry under possession

winstxnhdw commented:
I won’t have time to look at this until March, so @D1GQ please help with the testing in the meantime.

xAstroBoy commented:
@D1GQ hit me on discord for rapid feedback

D1GQ commented:
> @D1GQ hit me on discord for rapid feedback

Do you mind add me, is the same as my name here.

winstxnhdw commented:
Is unpossession not automatically going back to the player for you?

xAstroBoy commented:
@winstxnhdw yes, possession is not  going back to player for me, i fixed it.

winstxnhdw commented on lc-hax/Scripts/Modules/Possession/Controllers/BaboonHawkController.cs:
```diff
@@ -8,6 +9,30 @@ enum BaboonState {
+
+    public void OnPossess(BaboonBirdAI enemyInstance) {
+        if (BaboonBirdAI.baboonCampPosition != this.customCamp) return;
+        this.originalCamp = BaboonBirdAI.baboonCampPosition;
```
Why do we have to move the baboon's camp?

xAstroBoy replied:
To make it not drop stuff when we hold it

winstxnhdw commented on lc-hax/Scripts/Modules/Possession/Controllers/IEnemyController.cs:
```diff
@@ -29,9 +46,21 @@ internal interface IEnemyController<T> : IController where T : EnemyAI {
     string? GetPrimarySkillName(T enemyInstance) => null;
 
     string? GetSecondarySkillName(T enemyInstance) => null;
+    float? InteractRange(T enemyInstance) => 2.5f;
```
If most enemies have an interact range of `1.0f` why not just make that as default?

xAstroBoy replied:
only the small ones,  the big ones needs to be larger

winstxnhdw commented on lc-hax/Scripts/Modules/Possession/PossessionMod.cs:
```diff
@@ -190,15 +240,146 @@ internal sealed class PossessionMod : MonoBehaviour {
+    }
+
+
+    float InteractRange(EnemyAI enemy) {
+        return enemy is not EnemyAI enemyAI
+            ? 0
+            : this.EnemyControllers.TryGetValue(enemy.GetType(), out IController value)
+            ? value.InteractRange(enemyAI).GetValueOrDefault(2.5f)
+            : 2.5f;
+    }
+
+    float SprintMultiplier(EnemyAI enemy) {
+        return enemy is not EnemyAI enemyAI
+            ? 0
+            : this.EnemyControllers.TryGetValue(enemy.GetType(), out IController value)
+            ? value.SprintMultiplier(enemyAI).GetValueOrDefault(2.8f)
+            : 2.8f;
+    }
```
This is so cursed. You don't even need to do it. It's why we have default implementations in our interface.

winstxnhdw replied:
Nevermind, I am wrong here. I am not sure how I can do this better.

winstxnhdw commented on lc-hax/Scripts/Modules/Possession/PossessionMod.cs:
```diff
@@ -171,16 +211,26 @@ internal sealed class PossessionMod : MonoBehaviour {
-
     // Possesses the specified enemy
     internal void Possess(EnemyAI enemy) {
         this.Unpossess();
+        if (enemy.isEnemyDead) return;
```
Is there a reason for putting `unpossess` before this? I think this would make the player unpossess if they tried to possess a dead enemy.

xAstroBoy replied:
You right, fixed.

winstxnhdw commented:
Might be able to merge this by today or tomorrow, surprisingly. The PR aside from the `PossessionMod` script is mostly clean. Good job!

winstxnhdw commented:
@D1GQ is it just me or is the Jester colliders broken? I keep falling through the floor.

xAstroBoy commented:
> @D1GQ is it just me or is the Jester colliders broken? I keep falling through the floor.

that can be fixed by enlarging the collision 

> @D1GQ is it just me or is the Jester colliders broken? I keep falling through the floor.

yeah the charactercontroller collider is a little bit too small, can increase it a tad bit.

winstxnhdw commented:
![image](https://github.com/winstxnhdw/lc-hax/assets/56998716/f1396089-a710-439d-b535-f93221c85b6c)

Bracken seems to climb infinitely when it goes outside.

xAstroBoy commented:
> ![image](https://private-user-images.githubusercontent.com/56998716/305661713-f1396089-a710-439d-b535-f93221c85b6c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgxOTIyOTAsIm5iZiI6MTcwODE5MTk5MCwicGF0aCI6Ii81Njk5ODcxNi8zMDU2NjE3MTMtZjEzOTYwODktYTcxMC00MzlkLWI1MzUtZjkzMjIxYzg1YjZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE3VDE3NDYzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJhYzVmZjQ2MzRlOTAyNmQ4NzY5Mzg0MmZhZWZiMDRhZTdlZTcxYzliMGEzN2I1N2ViZjA2OWVhNDRlY2NiNzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KX3qHdX3FKNbO0OL1PgGA9U6NtYYJv-21tgAqc8L3A8)
> 
> Bracken seems to climb infinitely when it goes outside.

odd, that did not happen to me.

winstxnhdw commented:
Hold on, seems to be a mistake from me.

xAstroBoy commented on lc-hax/Scripts/Modules/Possession/PossessionMod.cs:
```diff
@@ -85,118 +95,265 @@ internal sealed class PossessionMod : MonoBehaviour {
+        characterMovement.SetPosition(enemy.transform.position);
+    }
+
+    void ResetInteractionCooldowns() {
+        this.TeleportCooldownRemaining = -DoorInteractionCooldown;
+        this.DoorCooldownRemaining = -TeleportDoorCooldown;
+    }
```
Is not a mistake, the negative cooldown is what makes it work

winstxnhdw replied:
```cs
        if (hit.collider.gameObject.TryGetComponent(out DoorLock doorLock) && this.DoorCooldownRemaining <= 0.0f) {
            this.OpenDoorAsEnemy(doorLock);
            this.DoorCooldownRemaining = PossessionMod.DoorInteractionCooldown;
            return;
        }
```

Okay, but it isn't necessary right? I thought it had to be negative or something would break while `0.0f` would be expected and work normally.

xAstroBoy replied:
That cooldown is supposed to go negative, once it hits any negative value it will work  and it will reset

winstxnhdw replied:
`<= 0.0f` means it’s true if it hits `0.0f` right?

xAstroBoy replied:
> `<= 0.0f` means it’s true if it hits `0.0f` right?
0.0 or below, yes

winstxnhdw commented:
@D1GQ I think we are done. Can you give us the okay?

winstxnhdw commented on lc-hax/Scripts/Modules/Possession/PossessionMod.cs:
Would this actually matter though? The next frame hasn’t even been generated yet so wouldn’t this operation be atomic?

xAstroBoy replied:
yeah because the charactercontroller will bug out

winstxnhdw replied:
I know this seems trivial but it’s quite important that we figure this out since we’ll probably forget and break things in the future when we move it around.

Have you tested it? Because when I was trying it out with the previous code, there wasn’t a problem for me. If it is really that unstable, we should add a comment saying so since this strays far from expected behaviour.

xAstroBoy replied:
> I know this seems trivial but it’s quite important that we figure this out since we’ll probably forget and break things in the future when we move it around.
> 
> Have you tested it? Because when I was trying it out with the previous code, there wasn’t a problem for me. If it is really that unstable, we should add a comment saying so since this strays far from expected behaviour.

no is not unstable, is just that .SetPosition() turns off the component, sets the transform.position then enables it back once it sets a position, that's the only way to force a character to teleport without screwing the controller.

xAstroBoy commented:
Before pushing, let me add lasso and spider , im as well adding a check for the secondary skill since is being spammed and is not required for all enemies

xAstroBoy commented:
@winstxnhdw 

i added as well OnSecondarySkillHold() and moved some enemies to use that instead and removed unneeded flags

Added Lasso, Thumper, Bees, Spider abilities .

removed IsSecondarySkill redundant flag.

Locked Nutcracker movement while is in sentry mode so it does not appear broken rotation and movement

Hopefully i fixed the centipede as well from soft locking.

winstxnhdw commented:
Are we good?

winstxnhdw commented:
Great job, let's merge after the CI has passed.
